### PR TITLE
WT-8812 Accept bucket location as a configuration parameter

### PR DESCRIPTION
**Pull request by bingx1**
A bucket name is not sufficient to uniquely identify an S3 bucket - the AWS region is also required. As part of this PR, this can now be passed as a config parameter to `customize_file_system`. [This list](http://sdk.amazonaws.com/cpp/api/LATEST/namespace_aws_1_1_region.html) contains all the valid values this parameter can take.

### CODE REVIEW

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 1]
```diff
@@ -382,12 +381,34 @@ S3CustomizeFileSystem(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session, con
...
            s3->wtApi, session, config, "prefix", &objPrefixConf)) == 0)
-        objPrefix = objPrefixConf.str;
+        objPrefix = std::string(objPrefixConf.str, objPrefixConf.len);
     else if (ret != WT_NOTFOUND) {
         std::cerr << "Error: customize_file_system: config parsing for object prefix";
         return (1);
     }
+    /*
+     * Configure the AWS Client configuration. The options that can be set are documented below:
+     * https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/client-config.html
```
Comment (sulabhM) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:391]: This URL is not for the CRT client, for instance, I could not find a reference to throughputTargetGbps or partSize. I reckon skip the url.

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 2]
```diff
@@ -382,12 +381,34 @@ S3CustomizeFileSystem(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session, con
...
+     */
+    Aws::S3Crt::ClientConfiguration awsConfig;
+    awsConfig.throughputTargetGbps = throughputTargetGbps;
+    awsConfig.partSize = partSize;
 
+    /*
+     * Get the AWS region to be used. If no region is provided, the region will default to the
+     * default region configured in the applicable AWS credentials. The allowable values for AWS
```
Comment (sulabhM) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:399]: Please remove the comment about defaulting to AWS credentials - the final version will take the credentials from the passed authToken (WT-8721). Let us make it mandatory to pass the location (similar to the bucket), absent which we will fail the customize file system call.

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 3]
```diff
@@ -382,12 +381,34 @@ S3CustomizeFileSystem(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session, con
...
+     * region are listed here in the AWS documentation:
+     * http://sdk.amazonaws.com/cpp/api/LATEST/namespace_aws_1_1_region.html
+     */
+    WT_CONFIG_ITEM regionConf;
+    std::string region;
+    if ((ret = s3->wtApi->config_get_string(s3->wtApi, session, config, "region", &regionConf)) ==
+      0)
+        awsConfig.region = std::string(regionConf.str, regionConf.len);
+    else if (ret != WT_NOTFOUND) {
```
Comment (sulabhM) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:408]: As mentioned above, I think it should be mandatory to have a location.

**Review by sulabhM — CHANGES_REQUESTED** [review]

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 3 above)
Comment (bingx1) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:404]: What should `customize_file_system` return if no region is provided? Should it to be 1? Or EINVAL like on line 372?

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 4]
```diff
@@ -382,12 +381,33 @@ S3CustomizeFileSystem(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session, con
...
+        std::cerr << "Error: customize_file_system: config parsing for AWS region";
+        return (1);
+    } else {
```
Comment (bingx1) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:407]: @sulabhM I added another clause to the if-statement here to differentiate between the case when no region has been provided, and when `config_get_string` actually fails for some other reason. Is this appropriate? Or would it be better to collapse these two paths into a single statement?

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 5]
```diff
@@ -382,12 +381,33 @@ S3CustomizeFileSystem(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session, con
...
+    if ((ret = s3->wtApi->config_get_string(s3->wtApi, session, config, "region", &regionConf)) ==
+      0)
+        awsConfig.region = std::string(regionConf.str, regionConf.len);
+    else if (ret != WT_NOTFOUND) {
+        std::cerr << "Error: customize_file_system: config parsing for AWS region";
+        return (ret);
+    } else {
+        std::cerr << "Error: Region not specified" << std::endl;
+        return (EINVAL);
+    }
```
Comment (bingx1) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:410]: @sulabhM - I decided to separate these cases to make the error paths clearer. Is this useful? Or would it be better to combine these into a single conditional?

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 6]
```diff
@@ -87,7 +87,6 @@ struct S3_FILE_HANDLE {
 };
 
 /* Configuration variables for connecting to S3CrtClient. */
-const Aws::String region = Aws::Region::AP_SOUTHEAST_2;
 const double throughputTargetGbps = 5;
 const uint64_t partSize = 8 * 1024 * 1024; /* 8 MB. */
```
Comment (ruby-oujo) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:91]: Would it be useful to also put these into the config string?

Comment (sulabhM) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:91]: Probably, but we are short of time now, so we are putting only the ones that are really required. Once this phase of the project is over, we will likely re-evaluate this question.

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 5 above)
Comment (sulabhM) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:410]: This is good, we need these error paths separated out.

**Review by sulabhM — APPROVED** [review]

**Pull request merged**